### PR TITLE
Defend goal situation

### PR DESCRIPTION
**Pull request by Jason27chan**
## Description
A situation for defending goal. Uses standard defense (two submissive defenders and a goalie) two wing defenders, and a marking robot for the highest threat enemy.

## Associated Issue
https://github.com/RoboJackets/robocup-software/issues/1345

## Design Documents
N/A

### Test Case 1
1. Run defend goal play

Expected result: two submissive defenders, two wingers, and 1 mark

Please provide feedback thank you

### CODE REVIEW

File: soccer/gameplay/evaluation/opponent.py [hunk 1]
```diff
@@ -65,3 +66,137 @@ def get_closest_opponent(pos, direction_weight=0, excluded_robots=[]):
                 closest_dist = dist
 
     return closest_bot
+
```
Comment (Jason27chan) [inline soccer/gameplay/evaluation/opponent.py:69]: Moved this code from defense.py

**Comment by kylestach** [comment]
@Jason27chan Is this pending #1357 (Situational plays beta)?

**Review by HussainGynai — APPROVED** [review]
Please make pretty lines. Looks good.

File: soccer/gameplay/evaluation/opponent.py (hunk 1 above)
Comment (tjones320) [reply soccer/gameplay/evaluation/opponent.py:69]: Do we think this is the long term best place for these sorts of things? I just don't want this file to become huge

File: soccer/gameplay/plays/defense/defend_goal.py [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+        for i in range(len(main.their_robots())):
+            self.defended[i] = False
+
+        for i in range(self.num_wingers):
+            self.add_subbehavior(wing_defender.WingDefender(),
+                                 'winger' + str(i))
+
+    def execute_running(self):
+        for i in range(self.num_wingers):
+            bhvr = self.subbehavior_with_name('winger' + str(i))
```
Comment (tjones320) [inline soccer/gameplay/plays/defense/defend_goal.py:50]: Does this make a list?

File: soccer/gameplay/plays/defense/defend_goal.py [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+
+            # Control which robots are being defended by wing defender
+            for i in range(len(main.their_robots())):
+                bot = main.their_robots()[i]
+                if not eval_opp.is_marked(bot.pos) and not self.defended[i]:
+                    bhvr.mark_robot = bot
```
Comment (tjones320) [inline soccer/gameplay/plays/defense/defend_goal.py:56]: still a little confused about how exactly this works.

File: soccer/gameplay/tactics/positions/wing_defender.py [hunk 4]
```diff
@@ -27,11 +28,13 @@ def __init__(self,
         super().__init__(continuous=True)
 
         self._goalside_ratio = goalside_ratio  # Ratio of angle defending goal line versus attacker line
-        self._mark_robot = mark_robot  # Robot we are defending against
+        # Robot we are defending against
+        self._mark_robot = eval_opp.get_closest_opponent(robocup.Point(0, 0)) \
```
Comment (tjones320) [inline soccer/gameplay/tactics/positions/wing_defender.py:31]: This is weird. Curious how does this work?

**Review by tjones320 — APPROVED** [review]

File: soccer/gameplay/plays/defense/defend_goal.py (hunk 2 above)
Comment (Jason27chan) [reply soccer/gameplay/plays/defense/defend_goal.py:50]: No, this tells the wing defenders which robots to mark

File: soccer/gameplay/evaluation/opponent.py (hunk 1 above)
Comment (Jason27chan) [reply soccer/gameplay/evaluation/opponent.py:69]: It is better than it being in defense.py. this is a big file though. I'll think about splitting up the file a bit

File: soccer/gameplay/tactics/positions/wing_defender.py (hunk 4 above)
Comment (Jason27chan) [reply soccer/gameplay/tactics/positions/wing_defender.py:31]: If we don't assign a robot, then the wing defender marks the closest robot to our goal

File: soccer/gameplay/plays/defense/defend_goal.py (hunk 3 above)
Comment (Jason27chan) [reply soccer/gameplay/plays/defense/defend_goal.py:56]: This loop marks any robot that's not currently marked

File: soccer/gameplay/evaluation/opponent.py (hunk 1 above)
Comment (tjones320) [reply soccer/gameplay/evaluation/opponent.py:69]: I think having multiple files in this folder will probably be the best plan going forward. Still, don't know if we want an even better structure than that.

**Pull request merged**